### PR TITLE
DNM PGLog: store extra duplicate ops beyond the normal log entries

### PR DESCRIPTION
**Pull request by ivancich**
This helps us avoid replaying non-idempotent client operations when the pg log is very short, e.g. in an effort to force OSDs to use backfill rather than regular recovery. This can be advantageous to avoid blocking i/o to objects, at the cost of longer total time to become clean (since backfill requires scanning the objects to see what is missing).

### CODE REVIEW

**Comment by ivancich** [comment]
@gregsfortytwo is thinking about a better way to test the dup-ops functionality. Shouldn't merge until better testing complete.

File: src/osd/PGLog.cc [hunk 1]
```diff
@@ -399,15 +438,93 @@ void PGLog::merge_log(pg_info_t &oinfo, pg_log_t &olog, pg_shard_t fromosd,
 
     changed = true;
   }
-  
-  dout(10) << "merge_log result " << log << " " << missing << " changed=" << changed << dendl;
+
+  // now handle dups
+  merge_log_dups(olog, changed);
```
Comment (jdurgin) [inline src/osd/PGLog.cc]: should set dirty_dups here if (changed), so they are persisted when write_log_and_missing() is called

File: src/osd/PGLog.h [hunk 2]
```diff
@@ -974,7 +1042,12 @@ struct PGLog : DoutPrefixProvider {
       rollbacker,
       this);
   }
+
+  void merge_log_dups(pg_log_t& olog,
+		      bool& changed);
```
Comment (jdurgin) [inline src/osd/PGLog.h]: minor: olog could be const, and convention is to pass out parameters as pointers to make them explicit - since there's just one you could make changed the return value instead

File: src/osd/PGLog.cc [hunk 3]
```diff
@@ -399,15 +438,93 @@ void PGLog::merge_log(pg_info_t &oinfo, pg_log_t &olog, pg_shard_t fromosd,
...
+	changed = true;
+
+	auto log_tail_version = log.dups.back().version;
+
+#if 0
+	// make sure the logs overlap
+	assert(olog.dups.front().version <= log_tail_version);
+#endif
+
+	auto insert_cursor = log.dups.end();
```
Comment (jdurgin) [inline src/osd/PGLog.cc]: could use splice() to simplify these two loops combining the lists (I guess that's a reason to keep olog non-const)

File: src/osd/PGLog.h [hunk 4]
```diff
@@ -517,6 +581,7 @@ struct PGLog : DoutPrefixProvider {
       (dirty_from != eversion_t::max()) ||
       (writeout_from != eversion_t::max()) ||
       !(trimmed.empty()) ||
+      !(trimmed_dups.empty()) ||
```
Comment (jdurgin) [inline src/osd/PGLog.h]: should check `dirty_dups`

File: src/osd/PGLog.cc (hunk 3 above)
Comment (jdurgin) [inline src/osd/PGLog.cc]: nvm, probably not worth it since you need to index them anyway

File: src/test/osd/TestPGLog.cc [hunk 5]
```diff
@@ -2064,6 +2066,269 @@ TEST_F(PGLogTest, ErrorNotIndexedByObject) {
   EXPECT_EQ(del.reqid, entry->reqid);
 }
 
+class PGLogMergeDupsTest : public ::testing::Test, protected PGLog {
```
Comment (jdurgin) [inline src/test/osd/TestPGLog.cc]: looks like you've got extensive tests of merging - it'd be good to have some for the trimming behavior as well

**Comment by jdurgin** [comment]
looks quite good overall!

**Comment by jdurgin** [comment]
Looking at the cache tiering test failures with high rates of messenger fault injection, the ones on master are due to having short pg logs and easily exceeding the bounds of the log for dup detection (two writes are proxied to the base tier, they end up being resent and only the 2nd is still in the log).

This particular failure (dup detection within a single tier) is fixed by the longer dup ops in this branch.
The reason it still fails with your branch is the way dup detection information is transferred between tiers. When promoting an object via the copy-get op, the response includes limited dup detection information by reading up to the last 10 entries and including those version numbers in the 'extra_reqids' field - `git grep get_object_reqids` to see this. This does not include anything from the log_dup_entries, so when enough retries occur, a later op still in the pg log is included in the promotion, and thus detected as a dup, and an earlier one is not and performed again, out of order.

To fix this I think we'd need to add the hobject_t to each dup entry as well, so we can include the dup_entrys in the extra_reqids during promotion. This isn't ideal, since it bloats the dup information more, but I don't see a better way to fix it. tiers are entirely different pools, so there is no one-to-one relationship between pgs and their dup logs.

File: src/test/osd/TestPGLog.cc [hunk 6]
```diff
@@ -2064,6 +2066,269 @@ TEST_F(PGLogTest, ErrorNotIndexedByObject) {
...
+      create_dup_entry(16, 32)
+    };
+    return result;
+  }
+
+  void add_dups(uint a, uint b) {
+    log.dups.push_back(create_dup_entry(a, b));
+  }
+
+  void add_dups(std::vector<pg_log_dup_t> l) {
```
Comment (gregsfortytwo) [inline src/test/osd/TestPGLog.cc]: I realize it's just test code, but these should probably all be const refs instead of pass-by-value?

File: src/common/config_opts.h [hunk 7]
```diff
@@ -899,6 +899,7 @@ OPTION(osd_min_pg_log_entries, OPT_U32, 3000)  // number of entries to keep in t
 OPTION(osd_max_pg_log_entries, OPT_U32, 10000) // max entries, say when degraded, before we trim
 OPTION(osd_force_recovery_pg_log_entries_factor, OPT_FLOAT, 1.3) // max entries factor before force recovery
 OPTION(osd_pg_log_trim_min, OPT_U32, 100)
+OPTION(osd_pg_log_dup_track_versions, OPT_U32, 1000) // how many versions back to track operations that could be duplicated in the pg log
```
Comment (gregsfortytwo) [inline src/common/config_opts.h]: We should think through how this interacts with the full pg log entries a little more. With the defaults here, the 1000 dup ops are just pointless work since the pg log entries are always longer. I see a few choices:
1) Turn the pg log entries *way* down when the PG is clean; only keep them around for log-based recovery when the PG is degraded. Increase dup op size accordingly.
2) run without dup ops by default, and expect people turn them up when turning down the pg log sizes.
3) Switch the config options around so that we have a "minimum tail of dup op detection" and auto-tune it so that we keep around as many dup ops as needed for that (which might be none, if the log is long enough, or if we want to "remember" the last 2000 ops and are only maintaining the last 1000 log entries, we switch them when evicting or something).

File: src/osd/osd_types.h [hunk 8]
```diff
@@ -3392,7 +3397,54 @@ WRITE_CLASS_ENCODER(pg_log_entry_t)
 
 ostream& operator<<(ostream& out, const pg_log_entry_t& e);
 
+struct pg_log_dup_t {
+  osd_reqid_t reqid;  // caller+tid to uniquely identify request
+  eversion_t version;
+  version_t user_version; // the user version for this entry
+  int32_t return_code; // only stored for ERRORs for dup detection
+  std::shared_ptr<hobject_t> soid_ref; // since hobject_t's are big, share them
```
Comment (gregsfortytwo) [inline src/osd/osd_types.h]: I'm not so sure about storing these. I know you only did it because of the failing cache-tier tests, but it really inflates the pg_log_dup_t size (I suspect there aren't going to be many repeats in most cache-tier cases?).

Given the use cases we're targeting, I don't think it's reasonable to expect to run base or cache pools without a real pg log. How about we enforce that via reasonable minimums, instead of contorting our cool new memory-slimming data structures to support them? :)

File: src/test/osd/TestPGLog.cc (hunk 5 above)
Comment (gregsfortytwo) [inline src/test/osd/TestPGLog.cc]: 👍 on unit tests for trimming.

There also aren't any explicit tests that it actually reports duplicate ops. :o Is that somehow covered by existing unit tests?

File: src/common/config_opts.h (hunk 7 above)
Comment (jdurgin) [inline src/common/config_opts.h]: As implemented this is a trailing number of dups stored past the end of the pg log - I like the idea of (3) though - just making this new option control total dup storage (pg log + dup log), along with decreasing the default pg log length. This is effectively (1) and (3), since there's already dynamic adjustment of how much we trim (whether to the min or max) based on being degraded or in recovery/backfill. See `PrimaryLogPG::calc_trim_to()`.

Comment (jdurgin) [inline src/common/config_opts.h]: This is currently trailing dup entries past the end of the pg log. I like the idea of (3) - changing this option to be total dup + pg log entries, so it adjusts the dup size based on the pg log size. Then we could make it 3000 (the current min pg log length) and lower the default min pg log length to e.g. 1000 to save memory when not recovering/backfilling/degraded.

(1) is taken care of already by calc_trim_to() - the pg log is trimmed less (only to the max, 10000 by default) during recovery/degraded/backfill.

Comment (ivancich) [inline src/common/config_opts.h]: I've implemented and am testing #3. My first test had PGPrimaryLog catching an out-of-order execution, so I've added more logging to trim and will re-test.

File: src/osd/osd_types.h (hunk 8 above)
Comment (ivancich) [inline src/osd/osd_types.h]: I've removed the commit that adds and manages the hobject_t's in the dup log.

**Comment by gregsfortytwo** [comment]
@ivancich, this is failing tests. Not sure if they're valid failures or not, but there are also a bunch of build warnings in the unit tests that make me wary.

**Comment by jdurgin** [comment]
afraid I caused some conflicts with my peering-deletes branch that just merged as well

**Comment by ivancich** [comment]
Perhaps one final issue is the default value of the three key config options. Previously osd_min_pg_log_entries was 3000 and osd_max_pg_log_entries was 10000. The PR in its current form sets osd_min_pg_log_entries to 1500, osd_max_pg_log_entries to 5000, and osd_pg_log_dups_tracked to 10000.

**Comment by ivancich** [comment]
Following discussion in today's core stand-up, this PR now sets osd_min_pg_log_entries to 1500 (reduced), osd_max_pg_log_entries to 10000 (unchanged), and osd_pg_log_dups_tracked to 3000 (to detect dups up to 3000 as did code before this PR).

**Comment by kungf** [comment]
hi, ivancich, what's "avoid replaying non-idempotent client operations when the pg log is very short" mean？

**Comment by ivancich** [comment]
@kungf An idempotent operation is one that you can apply multiple times without altering the underlying data. For example, you can write the byte 0xFF at offset 20 once, twice, or more times and the result is the same.

A non-idempotent operation is one where the underlying data would be different if you applied the operation a second, third, or more times. An example would be the operation to append the byte 0xFF to the end of the data. The data is different if you do it once or twice.

In ceph we do not want to apply the same (i.e., duplicate) operation more than once in case the op is non-idempotent. Ceph has used the pg log to detect duplicate operations. But there's a use case for one customer where we want to shorten the pg log, and so we might not detect a duplicate operation. This PR tracks ops beyond the pg log in a separate structure, so duplicate ops can still be detected.

**Comment by kungf** [comment]
@ivancich thank you very much!
